modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/WritableSessionCache.java

```diff
@@ -144,11 +147,19 @@ public WritableSessionCache( ExecutionContext context,
         this.referrerChangesForRemovedNodes = new HashMap<NodeKey, ReferrerChanges>();
         this.binaryReferencesByNodeKey = new ConcurrentHashMap<NodeKey, Set<BinaryKey>>();
         this.txns = repositoryEnvironment.getTransactions();
+        this.nodeTypes = repositoryEnvironment.nodeTypes();
     }
 
     protected final void assertInSession( SessionNode node ) {
         assert this.changedNodes.get(node.getKey()) == node : "Node " + node.getKey() + " is not in this session";
     }
+    
+    protected NodeTypes nodeTypes() {
+        if (this.nodeTypes == null && workspaceCache().repositoryEnvironment() != null) {
+            this.nodeTypes = workspaceCache().repositoryEnvironment().nodeTypes();
+        }  
```

Comment:
it's always the same instance and it will always be the repository's node types.
the problem is that this field may be null while the repository is initializing (i.e. during the RunningState ctr) while also `session.save` calls are being performed.
So after a repository has started up this should never be null.